### PR TITLE
[generate] prepare `attention_mask` outside `forward` whenever possible

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

WIP

For better use of `generate` + `flex_attention`

### CODE REVIEW

**Comment by gante** [comment]
Closing -- flex attention needs specialized manipulation of the block mask
